### PR TITLE
game engine

### PR DESCRIPTION
**Pull request by RuslanNikolaev**

### CODE REVIEW

File: src/test/groovy/org/dormitory/autobotsoccub/EqualsVerifierShould.groovy [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package org.dormitory.autobotsoccub
+
+import nl.jqno.equalsverifier.EqualsVerifier
+import org.dormitory.autobotsoccub.engine.model.Game
+import org.dormitory.autobotsoccub.engine.model.GameData
+import spock.lang.Specification
+
+class EqualsVerifierShould extends Specification {
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/EqualsVerifierShould.groovy]: Test class name should describe sut/test intention, e.g.:
EqualsAndHashcodeShould."be met by domain classes",
EqualsAndHashcodeTest."verify contract"

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    def setup() {
+        game = Game.builder()
+                .goalKeeperA(GK_A)
+                .forwardA(FW_A)
+                .goalKeeperB(GK_B)
+                .forwardB(FW_B)
+                .build()
+    }
+
+    def "correctly started game"() {
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: It would be better to split this test into 2 separate and add 1 more case:
1. "successfully start new game"
_make sure the game could be started by **ALL** registered players_
2. "fail to start already active game"
_verify against registered/unregistered user_
3. "fail to start game by unknown player"

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        then:
+        newGame
+        !existGame
+    }
+
+    def "correctly executed game commands"() {
+        given:
+        sut.startGame(FW_A, game)
+        sut.metaClass.getCurrentGame << { Integer userId ->
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: can this be moved to `setupSpec()` method?

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            GameData activeGame = activeGames.get userId
+            GameData.builder()
+                    .game(activeGame.getGame())
+                    .scoreTable(activeGame.getScoreTable())
+                    .build()
+        }
+
+        //correctly count score && count auto score && changed position in the team
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: well written test does not need such comments

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        when:
+        sut.score FW_A
+        sut.autoScore FW_A
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: let's stay with Java's method invocation syntax
you should be consistent with your style

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        forwardAStats.scored == 1
+        forwardAStats.autoScored == 1
+        forwardAStats.position == GK
+
+        //correctly catch exception
+        when:
+        sut.score 999
+
+        then:
+        GameDoesNotExistException e = thrown()
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: simply:
`
then: thrown GameDoesNotExistException
`
_in this case it is ok to omit parentheses_

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        when:
+        sut.score FW_B
+        sut.revert FW_B
+        sut.autoScore FW_B
+        sut.revert FW_B
+        sut.revert FW_B //check the situation without event
+        def gameData = sut.stopGame FW_B
+
+        then:
+        gameData.game != null
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: simply
`then:`
`gameData.game`
`gameData.scoreTable`

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    def "correctly started game"() {
+        when:
+        def newGame = sut.startGame(FW_A, game)
+        def existGame = sut.startGame(FW_A, game)
+
+        then:
+        newGame
+        !existGame
+    }
+
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: Add corresponding tests for `stopGame()` method

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        //correctly catch exception
+        when:
+        sut.score 999
+
+        then:
+        GameDoesNotExistException e = thrown()
+        e.cause == null
+
+        //correctly reverted last event & stop game
+        when:
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: 1. "correctly count scored goal" (use dataTable to supply different scores (per user) that should be verified) 
2. "fail to score" (think about edge cases)

add corresponding tests for autoScore and revert

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+        then:
+        gameData.game != null
+        gameData.scoreTable != null
+        def forwardBStats = gameData.scoreTable.statsByUserId.get(FW_B)
+        forwardBStats.scored == 0
+        forwardBStats.autoScored == 0
+    }
+
+    def "tested game"() {
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: "correctly handle game"

File: src/test/groovy/org/dormitory/autobotsoccub/engine/model/MatchTeamShould.groovy [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package org.dormitory.autobotsoccub.engine.model
+
+import spock.lang.Specification
+
+class MatchTeamShould extends Specification {
+
+    def "correctly returned another team"() {
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/model/MatchTeamShould.groovy]: return~~ed~~

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        sut.score GK_B
+
+        def userStats = sut.stopGame(FW_A).scoreTable.statsByUserId
+        def forwardA = userStats.get(FW_A)
+        def goalkeeperA = userStats.get(GK_A)
+        def forwardB = userStats.get(FW_B)
+        def goalkeeperB = userStats.get(GK_B)
+
+        then:
+        forwardA.scored == 3    && forwardA.autoScored == 1     && forwardA.missed == 2
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: use dataTable to specify your expectations

**Comment by jnopnop** [comment]
we almost there 🤠

File: src/test/groovy/org/dormitory/autobotsoccub/engine/model/MatchTeamShould.groovy (hunk 11 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/model/MatchTeamShould.groovy]: fixed

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 12 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: I deleted this method, 'cause the engine needs to be more developed for valuable game simulation.

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 10 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: thank you. sorry for my english)

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 9 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: In this section of the tests, the "Revert" command is checked, the extreme case is leaned - cancel the last event when it was already canceled. (This is the only extreme case for this test case)

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 8 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: fixed

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 7 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: simplified

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 6 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: simplified

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+                .forwardA(FW_A)
+                .goalKeeperB(GK_B)
+                .forwardB(FW_B)
+                .build()
+    }
+
+    def "successfully start new game"() {
+        expect:
+        sut.startGame(FW_A, game)
+        sut.stopGame(FW_A)
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: you don't need actually to stop game, the same for test below
you just need **_separate_** tests for `stopGame()` method

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 5 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: Ok. I'm sorry, I could not resist )

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 4 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: I see, thanks

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 3 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: What's the point of this? We call this method only within the method

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 2 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: Okay, I did it. but apart from "fail to start game by unknown player".('cause the engine and the user pool - are different entities. If unknown player called the game start command - the problem is not at the engine level)

File: src/test/groovy/org/dormitory/autobotsoccub/EqualsVerifierShould.groovy (hunk 1 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/EqualsVerifierShould.groovy]: I see, thank

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 3 above)
Comment (jnopnop) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: yep, but the test is incorrect itself. in my prev comment (prev commit) I mentioned that this test should be split into smaller parts; again, this is _unit-test_
>>
1. "correctly count scored goal" (use dataTable to supply different scores (per user) that should be verified)
2. "fail to score" (think about edge cases)
**1 and 2 should test only** `score` **method**

add corresponding tests for `autoScore` and `revert`

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy (hunk 13 above)
Comment (RuslanNikolaev) [reply src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: I misunderstood you. Fixed

File: src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy [hunk 14]
```diff
@@ -2,33 +2,59 @@ package org.dormitory.autobotsoccub.engine
...
+    @Shared
     def unknown_user = 999
 
+    @Shared
     Game game
+    @Shared
     GameEngine sut = new GameEngine()
 
-    def setup() {
+    def setupSpec() {
```
Comment (jnopnop) [inline src/test/groovy/org/dormitory/autobotsoccub/engine/GameEngineShould.groovy]: That's it 👍
